### ISSUE TITLE
Performance degradation in 2.x release

### ISSUE THREAD

**Issue by jaraco** [cannatag/ldap3#238]
Since updating to the 2.x release, our test suite now takes 2-3 times longer ~140sec vs ~60sec to complete than running the test with ldap3 1.4.0, all other things equal. I don't know if this difference will affect our production use of it.

Here's an example invoking a single connection/query:

```
$ hg update ldap3-2
$ time PYTHONPATH=.eggs/ldap3-2.0.9-py3.6.egg:.eggs/pyasn1-0.1.9-py3.6.egg python -m yg.ldap.commands --username ygapp_ygldap --password $YGAPP_YG_LDAP_PASSWORD jason
...
PYTHONPATH=.eggs/ldap3-2.0.9-py3.6.egg:.eggs/pyasn1-0.1.9-py3.6.egg python3 -  1.21s user 0.51s system 42% cpu 4.068 total
$ hg update master
$ time PYTHONPATH=.eggs/ldap3-1.4.0-py3.6.egg:.eggs/pyasn1-0.1.9-py3.6.egg python -m yg.ldap.commands --username ygapp_ygldap --password $YGAPP_YG_LDAP_PASSWORD jason
...
PYTHONPATH=.eggs/ldap3-1.4.0-py3.6.egg:.eggs/pyasn1-0.1.9-py3.6.egg python3 -  0.76s user 0.12s system 47% cpu 1.835 total
```

The only differences in the code between master and ldap3-2 are the compatibility changes. As you can see, the same operation takes almost twice as long. And if I run a full test suite, where the ldap3 2.x changes dominate the behavior, the factor is more noticeable.

Are there any changes in version 2 that are expected to impact performance? Have you benchmarked the behavior in any other environments? I expect the inferred schema for attributes is probably more expensive. Do you have any other advice for improving performance?

Again, it's possible the degradation is a non-issue, but I wanted to report my findings here and get any advice you might have.

**Comment by cannatag** [comment]
Version 2 by defaults read the server schema. So if you open and close the connection it will perform a search on the schema, that can takes a relatively long time. This is because RFC states that the client must be aware of the server schema and must perform checks before sending data. You can try to define the server with get_info=NONE then execution  time should be similar to version 1.4.

I was thinking of caching the schema but this should not be permitted by the RFC.

Let me know if you get faster with get_info set to NONE.

**Comment by jaraco** [comment]
I've tried setting get_info=None, but that seems to have a small effect on the performance of the tests. Our test suite starts up the connection only twice, and I can see when it does it, the connection is established in under a second, so isn't enough to account for the 80 sec differential. I can see from the test results that setting get_info=None restores the previous behavior of not processing the attribute schema. And that would explain the disparity with ldap3 v1 as well.

I'm thinking my initial results above may have been influenced by a bad network connection, as now I'm seeing a difference of only 20% now when `get_info` is not None (56 vs 70 sec).

In summary, you've effectively identified the cause of the difference, and the difference isn't as extreme as I'd initially suspected and the original performance, and the original behavior can be restored with the indicated workaround.

Thanks for the help.
